retryablemessageprocessor.go

```diff
@@ -0,0 +1,60 @@
+package messageprocessor
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/kinesis"
+	"github.com/rs/xstats"
+)
+
+// MaxRetries is the maximum number of time's we'd like to retry processing a record before quitting
+const consumerRetriesExceeded = "kinesis.consumer_error.retries_exceeded"
+
+// RetryableMessageProcessor is a `MessageProcessor` decorator that re-attempts
+// processing of messages 'maxAttempts' number of times in case of failures
+// 'maxAttempts' is a configurable parameter which can be set by consumer of this lib
+// Exponential backoff has been implemented as a retry mechanism
+type RetryableMessageProcessor struct {
+	maxAttempts int
+	wrapped     MessageProcessor
+}
+
+// ProcessMessage invokes the wrapped `MessageProcessor`. Attempts retries using exponential backoff
+// if underlying 'MessageProcessor' returns an error.
+// If 'maxAttempts' are exceeded without successful processing, it emits a stat indicating the same
+func (t *RetryableMessageProcessor) ProcessMessage(ctx context.Context, record *kinesis.Record) MessageProcessError {
```

Comment:
question: other than emitting stats, are we doing something more with kinesis record we are not able to process?

Reply:
Nope. apart from emitting stats (on retry exhaustion), we don't do anything else

Reply:
we will end up in poison pill scenario https://aws.amazon.com/blogs/compute/new-aws-lambda-controls-for-stream-processing-and-asynchronous-invocations/

Reply:
Well, the max retries get us out of the poison pill scenario, though admittedly poorly, but that was all the original ticket asked for. We are going to follow up with another subtask to define some standard deadlettering behaviors, which can actually be wrapped between the retry `MessageProcessor` and the stats `MessageProcessor` implementation.

Reply:
i am not sure, if we are not doing anything with message won't it end up back on kinesis stream?

Reply:
check pointing is handled in application code - https://bitbucket.org/asecurityteam/product-team-jira-sync/src/99eb9c0c96ba806bccb1447581722cc20b9953d1/pkg/consumer/kinesis.go#lines-112